File: leetcode/graph/bfs/0127_word_ladder.cpp
// LC 127. Word Ladder | Hard
#include <string>
#include <vector>
#include <queue>
#include <unordered_set>
using namespace std;

class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> words(wordList.begin(), wordList.end());
        if (!words.count(endWord)) return 0;
        queue<pair<string,int>> q; q.push({beginWord, 1});
        unordered_set<string> vis{beginWord};
        while (!q.empty()) {
            auto [word, steps] = q.front(); q.pop();
            for (int i = 0; i < word.size(); i++) {
                string nxt = word;
                for (char c = 'a'; c <= 'z'; c++) {
                    nxt[i] = c;
                    if (nxt == endWord) return steps + 1;
                    if (words.count(nxt) && !vis.count(nxt)) { vis.insert(nxt); q.push({nxt, steps+1}); }
                }
            }
        }
        return 0;
    }
};

File: leetcode/graph/bfs/0127_word_ladder.py
# LC 127. Word Ladder | Hard
from collections import deque

class Solution:
    def ladderLength(self, beginWord, endWord, wordList):
        wordSet = set(wordList)
        if endWord not in wordSet: return 0
        q = deque([(beginWord, 1)]); visited = {beginWord}
        while q:
            word, steps = q.popleft()
            for i in range(len(word)):
                for c in 'abcdefghijklmnopqrstuvwxyz':
                    new_word = word[:i] + c + word[i+1:]
                    if new_word == endWord: return steps + 1
                    if new_word in wordSet and new_word not in visited:
                        visited.add(new_word); q.append((new_word, steps + 1))
        return 0
